### ISSUE TITLE
Legion Spy support for > 3D regions

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#933]
If you want to run Legion Spy on regions with more than 3 dimensions (e.g. 4D), then you currently get the following assertion:

```
NotImplementedError: Need support for >3 dimensions
```

It would be nice to add this support since not all codes can be run with only 3 dimensions.

**Comment by lightsighter** [comment]
I'm inclined to put this in backlog. Is this just something nice you want to have or do you think this is necessary? You should know it probably involves writing a code generator to emit code for handling all the rectangle subtraction cases which grows as O(3^N).

**Comment by elliottslaughter** [comment]
I made this issue because this feature was just re-requested by a new user. So it seems to be a recurring request.

Do we need a code generator? Since this is Python, can't we just make a slow, generic version? Slow would be better than non-existent, which is what we've got now. How much effort would that be?

**Comment by lightsighter** [comment]
> Do we need a code generator? Since this is Python, can't we just make a slow, generic version?

What I mean is that you need a code generator to emit the Python code for all the different cases. Specifically we need a code generator to emit this function:
https://gitlab.com/StanfordLegion/legion/-/blob/master/tools/legion_spy.py#L506-1884
For 4-D there are 81 cases and since it is 4-D it's very difficult to visualize it as a human, so you pretty much need to rely on some math and a code generator to get it right. This code generator isn't for speed; it is for correctness.

> Slow would be better than non-existent, which is what we've got now. How much effort would that be?

We used to have the slow version that did set subtraction by testing all points individually, but it was so slow that it would just look like Legion Spy was hanging for anything that wasn't a very simple toy program. I would much rather the answer be to not support it at all so the user at least understands what the cause of the problem is.

**Comment by elliottslaughter** [comment]
> This code generator isn't for speed; it is for correctness.

My point is, anything that can be written as a code generator can just be written outright in Python. So if you think the generator would be shorter/easier to write and reason about, I'd just do it in Python itself. Maybe a little more generic, maybe a little (or a lot) slower, but there's no reason in my opinion to generate code rather than just writing it. If the generator is simpler than writing the code out, then make the generator your code.

**Comment by lightsighter** [comment]
> So if you think the generator would be shorter/easier to write and reason about, I'd just do it in Python itself. 

I actually don't care at all about which language it is or how you write the generator. I care about how much time it takes to figure out the math to make the generator and then how long it takes to create a test suite to convince yourself that the code you are generating is actually correct. The hard part here is the computational geometry, not the engineering.

**Comment by elliottslaughter** [comment]
We agree on this point.

Do you have an estimate of how much work this is, or is this still an unknown?

**Comment by lightsighter** [comment]
I would estimate it's at least one week of effort on my part. Someone with more familiarity thinking in higher dimensional spaces could probably do it faster. There's lots of symmetries in the problem to be exploited, but they are non-obvious at first glance. It is a nice self-contained problem though that doesn't require any other knowledge about anything else in Legion or Legion Spy. The problem statement is simple: given two rectangles A and B in N-D, return a set containing the minimum number of rectangles needed to represent the result of A - B. The kicker is the `minimum` part which is needed for good asymptotic complexity properties of the higher-level code.

**Comment by streichler** [comment]
Isn't the minimum just one rectangle per (hyper)face on which A sticks out of B?  An algorithm that iterates over the dimensions, carving off 0, 1, or 2 rectangles that contain the surplus in that dimension and extend as far as needed in any later dimensions, should hit this minimum.

**Comment by lightsighter** [comment]
Do you think you can do it greedily like that? I feel like dimension order would be important to hitting the minimum. I haven't been able to convince myself that it's not. I also have problems visualizing things in >3D so it could be that I'm just blind in this regard.

**Comment by streichler** [comment]
Yes, I think so.  I'm fairly confident about the minimum, because if I take any two points that are outside of B in exactly one dimension (i.e. adjacent to a hyperface), their convex hull would include points within B so they cannot be in the same rectangle.

And I'm similarly confident that the greedy algorithm results in exactly one created rectangle per hyperface exceeded.  This does assume the case of A and B being completely disjoint (i.e. A - B == A) being detected already, but the comment in the code you pointed to suggests that it has.

**Comment by lightsighter** [comment]
Ok, I've pushed a change for this to the master branch. I've kept the current code in place for dimensions 1-3 because it is about 50-100% faster (measured using pypy not cypython) than the recursive algorithm on random rectangle pairs (who knows what it is on the kinds of rectangles we'll find in applications). Asymptotically they are the same, but the overheads of temporary objects and their garbage collection in Python cause significant constant factor overheads. Even doing the tail-call optimization only gets back about 2-5% of that difference, so most of it is in the temporary objects themselves. I did test the recursive algorithm though on about 10M random rectangle pairs each for dimensions 1-3 and they all seemed to work just fine. Here is the code if anyone wants to review it.
https://gitlab.com/StanfordLegion/legion/-/blob/master/tools/legion_spy.py#L1987-2058
I'm closing the issue. If anyone has a problem they can reopen it.
